proposals/ref-struct-interfaces.md

```diff
@@ -0,0 +1,195 @@
+
+### ref struct interfaces
+<a name="ref-struct-interface"></a>
+The language will allow for `ref struct` types to implement interfaces. The syntax and rules are the same as for normal `struct` with a few exceptions to account for the limitations of `ref struct` types. 
+
+The ability to implement interfaces does not impact the existing limitations against boxing `ref struct` instances. That means even if a `ref struct` implements a particular interface,  it cannot be directly cast to it as that represents a boxing action.
+
+```csharp
+ref struct File : IDisposable
+{
+    private SafeHandle _handle;
+    public void Dispose()
+    {
+        _handle.Dispose();
+    }
+}
+
+File f = ...;
+// Error: cannot box `ref struct` type `File`
+IDisposable d = f;
+```
+
+The ability to implement interfaces is only useful when combined with the ability for `ref struct` to participate in generic arguments (as [laid out later](#ref-struct-generic)).
+
+To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when the a `class` implements the interface. 
+
+Default interface methods pose a problem for `ref struct` as there are no protections against the default implementation boxing the `this` member. 
+
+```csharp
+interface I1
+{
+    void M()
+    {
+        // Error: both of these box if I1 is implemented by a ref struct
+        I1 local1 = this;
+        object local2 = this;
+    }
+}
+
+ref struct S = I1 { }
+```
+
+To handle this a `ref struct` will be forced to implement all members of an interface, even if they have default implementations. The runtime will also be updated to throw an exception if a default interface member is called on a `ref struct` type.
+
+Detailed Notes:
+- A `ref struct` can implement an interface
+- A `ref struct` cannot participate in default interface members
+- A `ref struct` cannot be cast to interfaces it implements as that is a boxing operation
+
+### ref struct Generic Parameters
+<a name="ref-struct-generic"></a>
+The language will allow for generic parameters to opt into supporting `ref struct` as arguments by using the `allow T : ref struct` syntax: 
+
+```csharp
+T Identity<T>(T p) 
+    allow T : ref struct
+    => p;
+
+// Okay
+Span<T> local = Identity(new Span<int>(new int[10]));
+```
+
+This is similar to a `where` in that it specifies the capabilities of the generic parameter. The difference is `where` limits the set of types that can fullfill a generic parameter while the behavior defined here expands the set of types. This is effectively an anti-constraint as it removes the implicit constraint that `ref struct` cannot satisfy a generic parameter. As such this is given a new syntax, `allow`, to make that clearer.
```

Comment:
```suggestion
This is similar to a `where` in that it specifies the capabilities of the generic parameter. The difference is `where` limits the set of types that can fulfill a generic parameter while the behavior defined here expands the set of types. This is effectively an anti-constraint as it removes the implicit constraint that `ref struct` cannot satisfy a generic parameter. As such this is given a new syntax, `allow`, to make that clearer.
```

Reply:
I thought that was wrong but VS code didn't highlight it so I went along with it.